modules/backlogs/lib/open_project/backlogs/patches/work_packages/update_ancestors/loader_patch.rb

```diff
@@ -32,8 +32,12 @@ def self.included(base)
   module InstanceMethods
     private
 
+    def selected_descendants_attributes
+      super + %i(remaining_hours)
+    end
+
     def selected_leaves_attributes
-      super + [:remaining_hours]
+      super + %i[remaining_hours derived_remaining_hours]
     end
```

Comment:
:yellow_circle: I understand why `:remaining_hours` must be added to the `selected_descendants_attributes` array, as it is used in the derivation computation. And I suppose you fixed a bug because previously the `:remaining_hours` attribute would not be fetched in the hash, and so the sum would be incorrect. It would be correct only with the leaf work package, as the leaf is a real work package instance and not a hash. And it was not caught by the tests because in `modules/backlogs/spec/services/work_packages/update_ancestors_service_spec.rb` the parent work package did not have any remaining hours. Thanksfully you fixed it when you added `let(:parent_remaining_hours) { 2.0 }` in the spec.

But for the `selected_leaves_attributes` array, I do not see why `:remaining_hours` and `:derived_remaining_hours` are needed: in `WorkPackages::UpdateAncestors::Loader` class, `selected_leaves_attributes` is called only when `related_of(wp, :leaves)` is called, which is called only when `leaves_of` is called. And `leaves_of` is called only in `WorkPackages::UpdateAncestors#aggregate_done_ratio`. As remaining hours are not needed for done ratio computation, I would try to remove the `selected_leaves_attributes` method from the patch, commit, and see if something breaks.

It looks like the methods `selected_descendants_attributes` and `selected_leaves_attributes` were swapped by mistake in the original implementation.

What do you think?

Reply:
Quick update: removing the `derived_remaining_hours` from the list breaks all the tests on `UpdateAncestorsService` due to differing values on the children work packages (`nil` vs `0.0`).

Reply:
@cbliard Probably. This area could seriously get some love, in order to make the code clearer and maybe less "smart".

Would you keep this blocked off or move forward and add a code maintenance task to go over the possible misnomer among other small improvements?

Reply:
As I said in my review comment, I see no blockers in you PR: it works. I just have some tiny remarks and this one is a tiny remark too. I think the code is good enough like this. Not perfect but good enough. I would merge and move forward.

Reply:
I tried removing the `selected_leaves_attributes` method and I did not experience any failures when running `bundle exec rspec -fd spec/services/work_packages/update_ancestors_service_spec.rb modules/backlogs/spec/services/work_packages/update_ancestors_service_spec.rb`. Which test was failing exactly in your case?

Reply:
IIRC All of them. >.< It seems it wasn't able to match the objects as it was comparing `nil`s to `0.0`.

But now that you mention it, I didn't remove the method from here, I just kept it as it was originally.

By looking at the code I get the feeling that maybe, it was intended to eventually (that's too many layers of uncertainty :P ) it was considered to make `remaining_time` part of the `done_ratio` calculation, but we never fully committed to the idea. Probably @ulferts, @wielinde or @oliverguenther would know.